elasticsearch/src/main/scala/akka/stream/alpakka/elasticsearch/ElasticsearchFlowStage.scala

```diff
@@ -22,64 +22,118 @@ import akka.NotUsed
+    IncomingIndexMessage(source, passThrough)
 
   // Java-api - with passThrough
-  def create[T, C](id: String, source: T, passThrough: C, version: Long): IncomingMessage[T, C] =
-    IncomingMessage(Option(id), source, passThrough, Option(version))
+  def create[T, C](id: String, source: T, passThrough: C): IncomingMessage[T, C] =
+    IncomingIndexMessage(id, source, passThrough)
+}
+
+object IncomingUpdateMessage {
+  // Apply method to use when not using passThrough
+  def apply[T](id: String, source: T): IncomingMessage[T, NotUsed] =
+    IncomingMessage(Some(id), Some(source), NotUsed, Update, None, None)
+
+  // Apply method to use when using passThrough
+  def apply[T, C](id: String, source: T, passThrough: C): IncomingMessage[T, C] =
+    IncomingMessage(Some(id), Some(source), passThrough, Update, None, None)
+
+  // Java-api - without passThrough
+  def create[T](id: String, source: T): IncomingMessage[T, NotUsed] =
+    IncomingUpdateMessage(id, source)
 
   // Java-api - with passThrough
-  def create[T, C](id: String, source: T, passThrough: C, version: Long, indexName: String): IncomingMessage[T, C] =
-    IncomingMessage(Option(id), source, passThrough, Option(version), Option(indexName))
+  def create[T, C](id: String, source: T, passThrough: C): IncomingMessage[T, C] =
+    IncomingUpdateMessage(id, source, passThrough)
+}
+
+object IncomingUpsertMessage {
+  // Apply method to use when not using passThrough
+  def apply[T](id: String, source: T): IncomingMessage[T, NotUsed] =
+    IncomingMessage(Some(id), Some(source), NotUsed, Upsert, None, None)
+
+  // Apply method to use when using passThrough
+  def apply[T, C](id: String, source: T, passThrough: C): IncomingMessage[T, C] =
+    IncomingMessage(Some(id), Some(source), passThrough, Upsert, None, None)
+
+  // Java-api - without passThrough
+  def create[T](id: String, source: T): IncomingMessage[T, NotUsed] =
+    IncomingUpdateMessage(id, source)
 
   // Java-api - with passThrough
-  def create[T, C](source: T, passThrough: C): IncomingMessage[T, C] =
-    IncomingMessage(None, source, passThrough)
+  def create[T, C](id: String, source: T, passThrough: C): IncomingMessage[T, C] =
+    IncomingUpdateMessage(id, source, passThrough)
 }
 
-final case class IncomingMessage[T, C](id: Option[String],
-                                       source: T,
-                                       passThrough: C,
-                                       version: Option[Long] = None,
-                                       indexName: Option[String] = None) {
+object IncomingDeleteMessage {
+  // Apply method to use when not using passThrough
+  def apply[T](id: String): IncomingMessage[T, NotUsed] =
+    IncomingMessage(Some(id), None, NotUsed, Delete, None, None)
+
+  // Apply method to use when using passThrough
+  def apply[T, C](id: String, passThrough: C): IncomingMessage[T, C] =
+    IncomingMessage(Some(id), None, passThrough, Delete, None, None)
```

Comment:
I think we should only have a constructor with `id`. Passthrough could be added by calling `withPassthrough` builder method.

Reply:
Same as below.